### DISCUSSION TITLE
How to access the files that located in the Resources folder in the .NET MAUI app.

### DISCUSSION THREAD

**Discussion by sheiksyedm** [dotnet/maui#7070] [Q&A]
In our custom control, we have to access the files that are located in the application **Resources** folder. Can you please guide me how to access/read Resources folder files in the .NET MAUI?

**Comment by hbraasch** [comment]
Maybe I'm stating the obvious but here is some sample code. File does not have to be in the [Resources] folder, just marked as a EmbeddedResource. Notice the path needs to be set appropriately (see the string in [GetStreamFromFile($"Data.Syllables.{filename}"))] below). In my case I created a root folder called [Data], then a subfolder called [Syllables]. If the file is in the resources folder, the string will just be [$"Resources.{filename}"]. Hope it helps.

```
private void LoadSyllables(string filename, ref List<Syllable> syllablesList)
        {
            using (Stream stream = GetStreamFromFile($"Data.Syllables.{filename}"))
            {
                if (stream == null)
                {
                    throw new ApplicationException($"Unable to read syllables for [{filename}]");
                }

                using (StreamReader reader = new StreamReader(stream))
                {
                    var syllablesJson = reader.ReadToEnd();
                    if (string.IsNullOrEmpty(syllablesJson))
                    {
                        throw new ApplicationException ($"No data inside syllables file [{filename}]");
                    }
                    syllablesList = JsonConvert.DeserializeObject<List<Syllable>>(syllablesJson);
                }
            }
        }

public static Stream GetStreamFromFile(string filename)
        {
            var assembly = typeof(App).GetTypeInfo().Assembly;
            var assemblyName = assembly.GetName().Name;

            var stream = assembly.GetManifestResourceStream($"{assemblyName}.{filename}");

            return stream;
        }
```

**Reply by sheiksyedm** [reply]
@hbraasch  Thank you for your suggestion. In my requirement, I will provide two options to load files as FromResource() and FromFile(). In the first one, the file is marked as EmbeddedResource and loads through the FromResource() method. The second option, I don't want to keep the file as EmbeddedResource and it directly refers to the file path. If users enter only a file name rather than an absolute path, it will access the file from the application Resources folder like "Resources.{filename}".

**Reply by project-x51** [reply]
This works until the control is in a different class library from the App. In this case the typeof(App) command does not work. I solved this by using reflection to find the App class within all loaded types. Hardly ideal but it works :) 

If the code is run on Windows you can just use the EntryAssembly. For Android it did not work. I have not tried yet for IOS and MacCatalyst but I expect this solution will work for all OS's as it is vanilla .NET.

Use the below code and ensure the "Embedded resource" type is set for the "Build Action" on the resource file. The ResourceID should be something like "{Namespace}.Resources.Images.{Filename}" but will depend where you decide to put the resource. If you're not sure of the ResourceID just embed the resource, uncomment the code to list the resource names and look through the list to find your resource's ID.

```
/// <summary>
/// Returns a stream to an embedded resource 
/// </summary>
/// <param name="ResourceId">The ID of the resource.</param>
/// <exception cref="Exception">Throws an exception if it cannot find the Maui App or the resource</exception>
public Stream Get_ResourceStream(string ResourceID)
{

    // Get the assembly that contains the embedded resource 
    // Assembly aAssembly = Assembly.GetEntryAssembly(); - Works for Windows App
    Assembly aAssembly = AppDomain.CurrentDomain.GetAssemblies()
                    .SelectMany(s => s.DefinedTypes
                    .Where(t => t.Name == "App"
                                && t.BaseType.Equals(typeof(Microsoft.Maui.Controls.Application))
                        )
                    ).FirstOrDefault()?.Assembly
                    ?? throw new Exception("Unable to find the Maui App type. This is required to load embedded resources.");

    // Useful for finding the resource names
    //string[] aNames = aAssembly.GetManifestResourceNames();

    try
    {
        // Attempt to open a stream to the embedded resource
        return aAssembly.GetManifestResourceStream(ResourceID);
    }
    catch (Exception e)
    {
        throw new Exception($"Unable to open a stream to the embedded resource. Check it has been embedded and has the correct ID including prefix. ResourceID: {ResourceID}", e);
    }

}
```

**Comment by tbayart** [comment]
Hi,
Found this unanswered question so here my finding.
A added a **Audio** subdirectory in **Resources** and put **hello.mp3** file inside this subfolder.
To get a stream to the file, i use this simple code :
`using var stream = await FileSystem.OpenAppPackageFileAsync("Resources/Audio/hello.mp3");`
